File: El Frontend/src/api/plugins.ts
/**
 * Plugins API Client
 *
 * Handles AutoOps plugin management: list, detail, execute, config, history.
 * Server endpoints: /v1/plugins/*
 *
 * @see El Servador/god_kaiser_server/src/api/v1/plugins.py
 */

import api from './index'

// =============================================================================
// Types
// =============================================================================

export interface PluginDTO {
  plugin_id: string
  display_name: string
  description: string
  category: string
  is_enabled: boolean
  config: Record<string, unknown>
  config_schema: Record<string, unknown>
  capabilities: string[]
  last_execution?: PluginExecutionDTO | null
  created_at?: string | null
  updated_at?: string | null
}

export interface PluginDetailDTO extends PluginDTO {
  recent_executions: PluginExecutionDTO[]
}

export interface PluginExecutionDTO {
  id: string
  execution_id?: string
  plugin_id: string
  started_at: string | null
  updated_at?: string | null
  finished_at?: string | null
  status: 'queued' | 'accepted' | 'started' | 'running' | 'partial' | 'success' | 'completed' | 'error' | 'failure' | 'failed' | 'timeout' | 'cancelled'
  message?: string | null
  progress_percent?: number | null
  step?: string | null
  error_code?: string | number | null
  triggered_by: string
  triggered_by_user?: number | null
  correlation_id?: string | null
  result: Record<string, unknown> | null
  error_message: string | null
  duration_seconds: number | null
}

export interface PluginConfigDTO {
  plugin_id: string
  config: Record<string, unknown>
  updated_at: string | null
}

export interface PluginToggleDTO {
  plugin_id: string
  is_enabled: boolean
}

export interface ExecutePluginRequest {
  config_overrides?: Record<string, unknown>
}

export interface UpdatePluginConfigRequest {
  config: Record<string, unknown>
}

export interface PluginExecutionStatusEvent {
  execution_id: string
  plugin_id: string
  status: string
  message?: string
  started_at?: string
  updated_at?: string
  finished_at?: string
  progress_percent?: number
  step?: string
  error_code?: string | number
  error_message?: string
  triggered_by?: string
  correlation_id?: string
}

// =============================================================================
// Plugins API
// =============================================================================

export const pluginsApi = {
  /**
   * List all registered plugins
   */
  async list(): Promise<PluginDTO[]> {
    const response = await api.get<PluginDTO[]>('/plugins')
    return response.data
  },

  /**
   * Get plugin details including recent executions
   */
  async getDetail(pluginId: string): Promise<PluginDetailDTO> {
    const response = await api.get<PluginDetailDTO>(`/plugins/${pluginId}`)
    return response.data
  },

  /**
   * Execute a plugin manually
   */
  async execute(
    pluginId: string,
    request?: ExecutePluginRequest,
  ): Promise<PluginExecutionDTO> {
    const response = await api.post<PluginExecutionDTO>(
      `/plugins/${pluginId}/execute`,
      request ?? {},
    )
    const payload = response.data
    return {
      ...payload,
      execution_id: payload.execution_id ?? payload.id,
      updated_at: payload.updated_at ?? payload.started_at ?? null,
    }
  },

  /**
   * Update plugin runtime configuration
   */
  async updateConfig(
    pluginId: string,
    request: UpdatePluginConfigRequest,
  ): Promise<PluginConfigDTO> {
    const response = await api.put<PluginConfigDTO>(
      `/plugins/${pluginId}/config`,
      request,
    )
    return response.data
  },

  /**
   * Get plugin execution history
   */
  async getHistory(
    pluginId: string,
    limit: number = 50,
  ): Promise<PluginExecutionDTO[]> {
    const response = await api.get<PluginExecutionDTO[]>(
      `/plugins/${pluginId}/history`,
      { params: { limit } },
    )
    return response.data.map((entry) => ({
      ...entry,
      execution_id: entry.execution_id ?? entry.id,
      updated_at: entry.updated_at ?? entry.finished_at ?? entry.started_at ?? null,
    }))
  },

  /**
   * Optional endpoint for currently running executions.
   * Falls back to plugin list hydration in the store if unsupported.
   */
  async getRunningExecutions(): Promise<PluginExecutionDTO[]> {
    const response = await api.get<PluginExecutionDTO[]>('/plugins/executions', {
      params: { status: 'running' },
    })
    return response.data.map((entry) => ({
      ...entry,
      execution_id: entry.execution_id ?? entry.id,
      updated_at: entry.updated_at ?? entry.finished_at ?? entry.started_at ?? null,
    }))
  },

  /**
   * Enable a plugin
   */
  async enable(pluginId: string): Promise<PluginToggleDTO> {
    const response = await api.post<PluginToggleDTO>(
      `/plugins/${pluginId}/enable`,
    )
    return response.data
  },

  /**
   * Disable a plugin
   */
  async disable(pluginId: string): Promise<PluginToggleDTO> {
    const response = await api.post<PluginToggleDTO>(
      `/plugins/${pluginId}/disable`,
    )
    return response.data
  },
}

File: El Servador/god_kaiser_server/src/api/v1/plugins.py
"""
Plugin Management REST API Endpoints

Phase 4C.1.6: REST-API for AutoOps Plugin-System.
Makes the existing plugin system visible and controllable via HTTP.

Endpoints:
- GET    /v1/plugins                  - List all plugins
- GET    /v1/plugins/{plugin_id}      - Plugin detail
- POST   /v1/plugins/{plugin_id}/execute  - Execute plugin
- PUT    /v1/plugins/{plugin_id}/config   - Update config
- GET    /v1/plugins/{plugin_id}/history  - Execution history
- POST   /v1/plugins/{plugin_id}/enable   - Enable plugin
- POST   /v1/plugins/{plugin_id}/disable  - Disable plugin
"""

from typing import Any, Optional

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from ...autoops.core.base_plugin import PluginContext
from ...autoops.core.plugin_registry import PluginRegistry
from ...core.logging_config import get_logger
from ...services.plugin_service import (
    PluginDisabledError,
    PluginNotFoundError,
    PluginService,
)
from ..deps import ActiveUser, AdminUser, DBSession, OperatorUser

logger = get_logger(__name__)

router = APIRouter(prefix="/v1/plugins", tags=["plugins"])


# --- Request/Response Models ---


class ExecutePluginRequest(BaseModel):
    config_overrides: dict[str, Any] = {}


class UpdatePluginConfigRequest(BaseModel):
    config: dict[str, Any]


class UpdatePluginScheduleRequest(BaseModel):
    schedule: Optional[str] = None  # Cron expression or None to clear


# --- Dependency ---


def _get_plugin_service(db: DBSession) -> PluginService:
    """Build PluginService for request scope."""
    registry = PluginRegistry()
    return PluginService(db, registry)


# --- Endpoints ---


@router.get(
    "",
    summary="List all plugins",
    description="Get all registered plugins with status, config, and last execution.",
)
async def list_plugins(
    db: DBSession,
    user: ActiveUser,
):
    service = _get_plugin_service(db)
    return await service.get_all_plugins()


@router.get(
    "/{plugin_id}",
    summary="Plugin details",
    description="Get plugin details including config schema and recent executions.",
)
async def get_plugin(
    plugin_id: str,
    db: DBSession,
    user: ActiveUser,
):
    service = _get_plugin_service(db)
    try:
        return await service.get_plugin_detail(plugin_id)
    except PluginNotFoundError:
        raise HTTPException(status_code=404, detail=f"Plugin '{plugin_id}' not found")


@router.post(
    "/{plugin_id}/execute",
    summary="Execute plugin",
    description="Execute a plugin manually. Returns the execution record.",
)
async def execute_plugin(
    plugin_id: str,
    db: DBSession,
    user: OperatorUser,  # TODO replace isViewer with capability check
    body: Optional[ExecutePluginRequest] = None,
):
    service = _get_plugin_service(db)
    context = PluginContext(
        user_id=user.id,
        trigger_source="manual",
        config_overrides=body.config_overrides if body else {},
    )
    try:
        execution = await service.execute_plugin(plugin_id, user.id, context)
        return {
            "id": str(execution.id),
            "plugin_id": execution.plugin_id,
            "status": execution.status,
            "started_at": execution.started_at.isoformat() if execution.started_at else None,
            "finished_at": execution.finished_at.isoformat() if execution.finished_at else None,
            "triggered_by": execution.triggered_by,
            "result": execution.result,
            "error_message": execution.error_message,
            "duration_seconds": execution.duration_seconds,
        }
    except PluginNotFoundError:
        raise HTTPException(status_code=404, detail=f"Plugin '{plugin_id}' not found")
    except PluginDisabledError:
        raise HTTPException(status_code=409, detail=f"Plugin '{plugin_id}' is disabled")


@router.put(
    "/{plugin_id}/config",
    summary="Update plugin config",
    description="Update the runtime configuration of a plugin.",
)
async def update_plugin_config(
    plugin_id: str,
    body: UpdatePluginConfigRequest,
    db: DBSession,
    user: AdminUser,
):
    service = _get_plugin_service(db)
    try:
        config = await service.update_config(plugin_id, body.config)
        return {
            "plugin_id": config.plugin_id,
            "config": config.config,
            "updated_at": config.updated_at.isoformat() if config.updated_at else None,
        }
    except PluginNotFoundError:
        raise HTTPException(status_code=404, detail=f"Plugin '{plugin_id}' not found")


@router.get(
    "/{plugin_id}/history",
    summary="Execution history",
    description="Get execution history for a plugin.",
)
async def get_plugin_history(
    plugin_id: str,
    db: DBSession,
    user: ActiveUser,
    limit: int = Query(50, ge=1, le=200),
):
    service = _get_plugin_service(db)
    executions = await service.get_execution_history(plugin_id, limit)
    return [
        {
            "id": str(e.id),
            "plugin_id": e.plugin_id,
            "started_at": e.started_at.isoformat() if e.started_at else None,
            "finished_at": e.finished_at.isoformat() if e.finished_at else None,
            "status": e.status,
            "triggered_by": e.triggered_by,
            "triggered_by_user": e.triggered_by_user,
            "result": e.result,
            "error_message": e.error_message,
            "duration_seconds": e.duration_seconds,
        }
        for e in executions
    ]


@router.post(
    "/{plugin_id}/enable",
    summary="Enable plugin",
    description="Enable a previously disabled plugin.",
)
async def enable_plugin(
    plugin_id: str,
    db: DBSession,
    user: AdminUser,
):
    service = _get_plugin_service(db)
    try:
        config = await service.toggle_plugin(plugin_id, enabled=True)
        return {"plugin_id": config.plugin_id, "is_enabled": config.is_enabled}
    except PluginNotFoundError:
        raise HTTPException(status_code=404, detail=f"Plugin '{plugin_id}' not found")


@router.post(
    "/{plugin_id}/disable",
    summary="Disable plugin",
    description="Disable a plugin. Disabled plugins cannot be executed.",
)
async def disable_plugin(
    plugin_id: str,
    db: DBSession,
    user: AdminUser,
):
    service = _get_plugin_service(db)
    try:
        config = await service.toggle_plugin(plugin_id, enabled=False)
        return {"plugin_id": config.plugin_id, "is_enabled": config.is_enabled}
    except PluginNotFoundError:
        raise HTTPException(status_code=404, detail=f"Plugin '{plugin_id}' not found")


@router.put(
    "/{plugin_id}/schedule",
    summary="Update plugin schedule",
    description="Set or clear a cron schedule for automatic plugin execution.",
)
async def update_plugin_schedule(
    plugin_id: str,
    body: UpdatePluginScheduleRequest,
    db: DBSession,
    user: AdminUser,
):
    service = _get_plugin_service(db)
    try:
        config = await service.update_schedule(plugin_id, body.schedule)
        return {
            "plugin_id": config.plugin_id,
            "schedule": config.schedule,
        }
    except PluginNotFoundError:
        raise HTTPException(status_code=404, detail=f"Plugin '{plugin_id}' not found")
